flax/errors.py

```diff
@@ -297,5 +323,263 @@ def __call__(self, x):
+
+      def __call__(self, x):
+        conv = nn.Conv(features=3, kernel_size=3)
+
+    Foo().init(random.PRNGKey(0), jnp.zeros((1,)))
+
+  Note that this error is also thrown if you partially defined a Module inside
+  setup::
+
+    class Foo(nn.Module):
+      def setup(self):
+        self.conv = functools.partial(nn.Conv, features=3)
+
+      def __call__(self, x):
+        x = self.conv(kernel_size=4)(x)
+        return x
+
+    Foo().init(random.PRNGKey(0), jnp.zeros((1,)))
+
+  In this case, ``self.conv(kernel_size=4)`` inside ``__call__`` constructs a 
+  ``Conv``, which, upon creation, checks where inside the parent Module ``Foo``
+  it was created, and realize this was outside of setup.
+  """
+  def __init__(self, cls):
+    super().__init__(f'Submodule {cls} must be defined in `setup()` or in a '
+                     'method wrapped in `@compact`')
+
+
+class ModuleAttributeNotFoundError(FlaxError):
+  """
+  This error occurs when trying to access a module attribute that does not
+  exist. For instance, the following code will throw the error because attribute
+  ``unknown`` is not defined::
+
+    class Foo(nn.Module):
+      @nn.compact
+      def __call__(self, x):
+        y = self.unknown  # <--- ERROR
+        return self.embed(x)
+
+    foo = Foo()
+    variables = foo.init(rand.PRNGKey(0), jnp.ones((1,)))
+    foo.apply(variables, jnp.ones((1,))
+
+  The error also occurs when trying to call an unbound non-compact module 
+  method::
+
+    class Foo(nn.Module):
+      def setup(self):
+        self.dense = nn.Dense(features=4)
+
+      def __call__(self, x):
+        return self.dense(x)
+    
+    foo = Foo()
+    foo(np.arange(5))
+
+  The reason is that by directly calling ``foo(np.arange(5))``, ``setup()`` is
+  not called first, so ``self.dense`` is undefined. To resolve this, the module
+  should be called using :meth:`Module.apply() <flax.linen.Module.apply>` 
+  instead, which will internally try to call ``setup``::
+
+    foo = Foo()
+    variables = foo.init(rand.PRNGKey(0), np.arange(5))
+    foo.apply(variables, np.arange(5))
+  """
+  def __init__(self, module_name, attr_name):
+    super().__init__(f'"{module_name}" object has no attribute "{attr_name}"')
+
+
+class SetAttributeInModuleSetupError(FlaxError):
+  """
+  You are not allowed to modify Module class attributes in
+  :meth:`Module.setup() <flax.linen.Module.setup>`::
+
+    class Foo(nn.Module):
+      features: int = 6
+
+      def setup(self):
+        self.features = 3  # <-- ERROR
+
+      def __call__(self, x):
+        return nn.Dense(self.features)(x)
+
+    variables = SomeModule().init(random.PRNGKey(0), jnp.ones((1, )))
+
+  Instead, these attributes should be set when initializing the Module::
+
+    class Foo(nn.Module):
+      features: int = 6
+
+      @nn.compact
+      def __call__(self, x):
+        return nn.Dense(self.features)(x)
+
+    variables = SomeModule(features=3).init(random.PRNGKey(0), jnp.ones((1, )))
+        
```

Comment:
Maybe a TODO to link to a design note explaining why it's necessary for modules to stay frozen (otherwise we can't safely clone them, which we use for lifted transformations)